Gutenberg: Add a Related Posts block

tyxla opened the pull request:
Currently, the Related Posts functionality works as a singleton - multiple instances aren't expected, and overriding settings inline isn't easy because of the way we currently fetch settings and build the markup. We are keeping the singleton behavior, but we are now adding a new way to render related posts - the Gutenberg block.

This PR registers a Related Posts Gutenberg block uses the same markup for the Gutenberg block, without reusing the existing functionality to build the markup. This allows us to reuse most of the existing functionality provided by the module.

If you wish to learn more about the block UI, you can have a look at https://github.com/Automattic/wp-calypso/pull/26530.

#### Changes proposed in this Pull Request:

* Register a new Gutenberg block for Related Posts.
* Use the existing shortcode/widget markup to create block server rendering from scratch.
* Prevent the customizer from modifying the block markup in the frontend.
* Reuse the excluded posts functionality, to prevent from going in circle, back and forth between the same 2 related posts.
* Prevent from displaying legacy related posts when there is a Related Posts block in the current post content
* Reorder some of the module JS to prevent it from inserting related posts in the dedicated wrapper when it shouldn't

#### Testing instructions:

* Use one of your Jetpack sites.
  * Ideally, test this on a site that has some posts already, in order to be able to use the already indexed posts without the need to create related posts and wait for indexing.
  * If you don't have a local site you can test on, spin up a new Jurassic Ninja site with Jetpack, Gutenberg and the Calypso built block. You might want to try with a non-shortlived site to be able to test the front end. You can use the following link to spin up a site with all this: https://jurassic.ninja/create/?gutenpack&jetpack-beta&branch=add/jetpack-related-posts-block-server-reg 
* Connect the site to WP.com, and activate all recommended features.
* If you don't wish to create a bunch of posts yourself, [import the example theme unit test content](https://raw.githubusercontent.com/WPTRT/theme-unit-test/master/themeunittestdata.wordpress.xml).
* Write a post, insert a Related Posts block. 
* Play the block settings, verify it looks and works well both in wp-admin and in the frontend of your site.
* Note that it will take some time for related posts to be indexed, recognized and displayed in the front end. Also, some minimum of posts need to be created - creating 10 posts should be enough.
* Verify that customizer on a post with related posts block doesn't touch the markup of the block.
* Verify that when navigating from post A to its related post B, post B's related posts output doesn't contain a link to A.

#### Proposed changelog entry for your changes:

* Introduced a new Related Posts block for Gutenberg.

tyxla committed:
Add server rendering for Related Posts block

tyxla committed:
Setting block attributes properly

tyxla committed:
Use a custom block rendering markup

tyxla committed:
Add some semicolons

keoshi reviewed (commented):
This is looking great and runs smoothly!

Here's a couple of notes from my testing:

- The block settings should follow the order of the resulting visual: `Display thumbnails`, `Display date`, `Display context (category or tag)`.
- The input for `Number of posts` accepts number above 3. Try to enter 5 for instance, and check that it attributes the class `columns-5` to the element in grid view mode.

tyxla commented:
Thanks @keoshi, feedback has been addressed in https://github.com/Automattic/wp-calypso/pull/26530/commits/cf622a21ef5ad94dc2e239e0858979544574966e and https://github.com/Automattic/wp-calypso/pull/26530/commits/e1fb9e4b85f95303b6c0eccd5231e9c8eecd3254 and this branch has been updated accordingly to allow changes to be in the new JN sites that we spin up for testing.

tyxla commented:
Rebased, removed the built block and updated description to reflect the fact that we can now test without having to include the built block in the Jetpack PR. Thanks Osk❤️

tyxla committed:
Remove headline from server side rendering

tyxla commented:
This one is ready for review.

mapk commented:
Hey @tyxla! Looking great! I only have a few suggested design changes.

### **1. New icon for centering content**
Right now it appears we're using the Gutenberg image width default icon. Let's switch this to use the `outline-format_align_center` icon instead from Material icons. It looks like this:

![screen shot 2018-10-17 at 1 08 04 pm](https://user-images.githubusercontent.com/617986/47113372-b782a380-d20d-11e8-8bab-90a1da6f9f9c.png)

Download it here: https://material.io/tools/icons/?icon=format_align_center&style=outline

### **2. Let's turn the category OFF by default.**

### **3. When the category is switched ON, it should be styled differently.** 
Here's an example...

![example](https://user-images.githubusercontent.com/617986/47114030-8d31e580-d20f-11e8-8bd5-d36fae60f9ad.png)

The CSS I wrote for this change looked like this:
```
.wp-block-a8c-related-posts__preview-post > p {
    color: #6c7781;
    font-size: 12px;
    margin: 0;
    text-transform: uppercase;
}
```

MichaelArestad commented:
> The CSS I wrote for this change looked like this:

I would just be cool if it was that size and not all caps. In Calypso, we found all caps not particularly easy to read. It also might do weird things depending on the themes font (but that goes for just about everything we make). Either way, I think those are great improvements.

sirreal commented:
@mapk some of that feedback seems to be oriented at the editor, this PR is for rendering the saved related posts block on the frontend of the site. Here's a screenshot with a standard WordPress setup (twentyseventeen theme):

![cars](https://user-images.githubusercontent.com/841763/47147078-b5801980-d2cd-11e8-8f19-0da3b5324bf2.png)


We can get a PR ready for those editor tweaks 👍

jeherve commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+
+		ob_start();
+		?>
+		<div id="jp-relatedposts" class="jp-relatedposts" style="display: block;">
```
I am seeing 2 duplicate sets of 3 related posts on my posts. 

<img width="554" alt="screenshot 2018-10-18 at 13 05 25" src="https://user-images.githubusercontent.com/426388/47150318-828e5380-d2d6-11e8-8eaa-4c1f8eacd542.png">

One is added via the Gutenberg block, the other is added via the regular Related Posts and gets injected right here because it finds the `jp-relatedposts` ID in the post content:
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/related-posts.js#L231

What should the expected behaviour be here? Should we hide the posts at the bottom of the page when a block has been added in the post content, like we do with the Related Posts shortcode?
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/jetpack-related-posts.php#L171

Or should we display the same Related Posts twice, once in the post content, once below it? If we choose to go that way, we should make sure the Gutenberg Related Posts don't use the same ID.

sirreal replied:
If we're going to allow multiple blocks like this, we can't use a static ID for them, that would be invalid HTML. IDs must be globally unique in the document.

See https://github.com/Automattic/wp-calypso/pull/27927#issuecomment-431044923

That would solve the some of the problems @jeherve [mentions](https://github.com/Automattic/jetpack/pull/10132/files#r226263162), but unfortunately requires a fair amount of styling work, as this is used for lots of styling: https://github.com/Automattic/jetpack/search?l=CSS&q=jp-relatedposts

tyxla replied:
So, if we want to reuse most of the existing functionality (which I guess we do), we should keep using the same ID, but also to have a maximum of 1 of those blocks per post. So I've suggested a PR to add this limitation to the block: 

https://github.com/Automattic/wp-calypso/pull/29415

Additionally, some small modifications are necessary so the related posts script will be able to properly work, but not load additional posts on top of ours. I've done that here: https://github.com/Automattic/jetpack/pull/10132/commits/73b1f29b3c01fa2845b75b14070eaf264cbbabeb

sirreal committed:
Default to off for displayContext

Requires Calypso-side change:

https://github.com/Automattic/wp-calypso/pull/27924

jeherve commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+							<a
+								class="jp-relatedposts-post-a"
+								href="<?php echo esc_url( $related_post['url'] ); ?>"
+								title="<?php echo esc_attr( $related_post['title'] ); ?>"
```
The existing Related Posts display the excerpt here as well.
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/related-posts.js#L58

Should we do the same here?

tyxla replied:
Good idea, done in: https://github.com/Automattic/jetpack/pull/10132/commits/809e5533e303be4594c178a2f9933b770c1bfd96

jeherve reviewed (dismissed):
I left a few notes about things that could be changed. 

I would also recommend looking at the WPCS reviews for the code you've added; there were a few recommendations you could follow to make sure your code respects the coding standards.

All in all, is there a reason you did not follow the existing Related Posts implementation that adds the posts via JavaScript?
https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/related-posts.js
It seems like it would have been easier to make sure the posts look like the existing Related Posts.

sirreal commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+							<a
+								class="jp-relatedposts-post-a"
+								href="<?php echo esc_url( $related_post['url'] ); ?>"
+								title="<?php echo esc_attr( $related_post['title'] ); ?>"
```
`title` and `excerpt` should both be included with an empty line between:

https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/related-posts.js#L56-L59

tyxla replied:
Done in https://github.com/Automattic/jetpack/pull/10132/commits/809e5533e303be4594c178a2f9933b770c1bfd96

sirreal commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+						<p class="jp-relatedposts-post-excerpt"><?php echo esc_html( $related_post['excerpt'] ); ?></p>
+
+						<?php if ( $block_attributes['show_date'] ): ?>
+							<p class="jp-relatedposts-post-date" style="display: block;">
```
There are so many missed opportunities for structured data here.

Just flagging this because I know it's something @davidakennedy was thinking about.

pafL3P-2b-p2

tyxla replied:
Good idea! Let's consider this, but in a different, separate effort which would target both the block and the legacy Related Posts markup.

sirreal reviewed (dismissed):
Nice start here, there's a lot going on and this must have been difficult to get together!

I'll start by saying I'm not familiar with related posts in general, so there was a lot of discovery for me as I worked through this. I left a lot of feedback.

---

The old, non-Gutenberg version would inject an HTML skeleton, then some frontend JS injects related posts into that and makes the necessary modifications:

```html
<div id="jp-relatedposts" class="jp-relatedposts">
	<h3 class="jp-relatedposts-headline"><em>Related</em></h3>
</div>
```

The id `#jp-relatedposts` is used by themes for styling, notably twentysixteen and twentyseventee, as well as by the related posts JS to target the element and inject some content. That means we need to keep the ID, but then we have some issues:

- Duplicate IDs (HTML no-no)
- The related posts JS targets the first element it finds, so there's an empty (and hidden) related posts skeleten, and the related posts block contains its original related posts and also the injected related posts 😞 It's in the wrong place.

Related posts seems like a 1-per-page thing, so we might consider:

- When a block is processed, remove the filter that's injecting the related posts skeleton. This should save the duplicate IDs.
  https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/jetpack-related-posts.php#L1444
- Dequeue the related posts script
  https://github.com/Automattic/jetpack/blob/bb884c04b32a57e3053f14c56e2f4e0138aef185/modules/related-posts/jetpack-related-posts.php#L1456-L1464

---

Given that we're producing a static ID in every block, it may make sense not to allow it to be inserted more than once: https://github.com/Automattic/wp-calypso/pull/27927

---

A final bit of strangeness between non-Gutenberg and Gutenberg related posts is the customizer.

Some of this is improved by not including the non-G related post bits when a Gutenberg related posts block has been detected as above.

Still, the customizer will attempt to "preview" changes to Gutenberg related posts as you modify the related posts settings, which isn't right and makes for a misleading preview 😞 

We need to add something to distinguish between Gutenberg-generated and non-G related posts. Something like this fixes the customizer experience to only target the non-G versions.

```patch
diff --git a/modules/related-posts/class.related-posts-customize.php b/modules/related-posts/class.related-posts-customize.php
index 8349b5300..7cf203afb 100644
--- a/modules/related-posts/class.related-posts-customize.php
+++ b/modules/related-posts/class.related-posts-customize.php
@@ -95,7 +95,7 @@ class Jetpack_Related_Posts_Customize {
 		// If selective refresh is available, implement it.
 		if ( isset( $wp_customize->selective_refresh ) ) {
 			$wp_customize->selective_refresh->add_partial( "$this->prefix", array(
-				'selector'            => '.jp-relatedposts',
+				'selector'            => '.jp-relatedposts:not([data-generated-by="gutenberg"])',
 				'settings'            => $selective_options,
 				'render_callback'     => __CLASS__ . '::render_callback',
 				'container_inclusive' => false,
diff --git a/modules/related-posts/jetpack-related-posts.php b/modules/related-posts/jetpack-related-posts.php
index 565ed4dee..86fc64b96 100644
--- a/modules/related-posts/jetpack-related-posts.php
+++ b/modules/related-posts/jetpack-related-posts.php
@@ -297,7 +297,7 @@ EOT;
 
 		ob_start();
 		?>
-		<div id="jp-relatedposts" class="jp-relatedposts" style="display: block;">
+		<div id="jp-relatedposts" class="jp-relatedposts" style="display: block;" data-generated-by="gutenberg">
 			<div class="jp-relatedposts-items jp-relatedposts-items-visual jp-relatedposts-<?php echo esc_attr( $block_attributes['layout'] ); ?>">
 				<?php foreach ( $related_posts as $index => $related_post ): 
 					$classes = array_filter( array(
diff --git a/modules/related-posts/related-posts.js b/modules/related-posts/related-posts.js
index ac2eb0d4f..7adde3b73 100644
--- a/modules/related-posts/related-posts.js
+++ b/modules/related-posts/related-posts.js
@@ -201,7 +201,7 @@
 		jprp.cleanupTrackedUrl();
 
 		var endpointURL = jprp.getEndpointURL(),
-			$relatedPosts = $( '#jp-relatedposts' );
+			$relatedPosts = $( '#jp-relatedposts:not([data-generated-by="gutenberg"])' );
 
 		$.getJSON( endpointURL, function( response ) {
 			if ( 0 === response.items.length || 0 === $relatedPosts.length ) {
```

MichaelArestad commented:
> @mapk some of that feedback seems to be oriented at the editor, this PR is largely for rendering the saved related posts block on the frontend of the site. Here's a screenshot with a standard WordPress setup (twentyseventeen theme):

@sirreal Huh! That's a pretty big change. The Gutenberg block in the editor should represent what will be on the front end (with the exception of what a theme adds to the styles).

Are old related posts styles being applied on the front end?

tyxla commented:
Thanks for the feedback @mapk!

> 1. New icon for centering content

Well, the icon comes from the Gutenberg core - we're using the `BlockAlignmentToolbar` component, so if you wish to change this, you might want to suggest that upstream.

> 2. Let's turn the category OFF by default.

Addressed by @sirreal in https://github.com/Automattic/jetpack/pull/10132/commits/0a2e374f3a67a9a26a0910f75d1ea56664b81a0c and https://github.com/Automattic/wp-calypso/pull/27924

> 3. When the category is switched ON, it should be styled differently.

Thanks for this! I've taken your CSS and @michaelarestad's suggestion for non-capitalized context and suggested a PR for this here: https://github.com/Automattic/wp-calypso/pull/27934

tyxla commented:
> Are old related posts styles being applied on the front end?

Currently, yes, and I think this is good for backwards compatibility - this was one of the reasons to use the old selectors and markup in the frontend.

mapk commented:
> Well, the icon comes from the Gutenberg core - we're using the `BlockAlignmentToolbar` component, so if you wish to change this, you might want to suggest that upstream.

@tyxla, I get that we're using the `BlockAlignmentToolbar` component, but in our block, this seems to be used as a _content_ alignment tool rather than a _block_ aligning tool. Is that a correct assumption? If this is the case, I don't think we need to file a PR upstream, unless we need to propose a "content" alignment component.

sirreal commented:
> I get that we're using the BlockAlignmentToolbar component, but in our block, this seems to be used as a content alignment tool rather than a block aligning tool. Is that a correct assumption?

It's a bit trickier than that, we're using the `BlockAlignmentToolbar` to align the content _and_ the block, which may not be exactly what it was intended for.

It's clearer when your theme supports wide/full alignment (one of the things the component does is check for theme support before showing the additional alignment options). Take a look at this:

<img width="828" alt="center-wide-full" src="https://user-images.githubusercontent.com/841763/47183335-0de40500-d328-11e8-8e21-3acef7ddf401.png">

The block/content alignment distinction may make sense. What if I want a wide block with the related posts content centered?

I threw together a PR to see what was involved, have a look: https://github.com/Automattic/wp-calypso/pull/27942

Note that the wide and full block alignments are lost in that PR.

tyxla committed:
Pass render callback to jetpack_register_block()

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -79,6 +79,11 @@ public function __construct( $blog_id_local, $blog_id_wpcom ) {
 		}
+
+		// Register Related Posts Gutenberg block
+		if ( function_exists( 'register_block_type' ) ) {
```
Even better - we're now using `jetpack_register_block()` that does all the checks and heavy lifting for us.

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -79,6 +79,11 @@ public function __construct( $blog_id_local, $blog_id_wpcom ) {
 		}
+
+		// Register Related Posts Gutenberg block
+		if ( function_exists( 'register_block_type' ) ) {
```
Even better - we're now using `jetpack_register_block()` that does all the checks and heavy lifting for us.

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+							</a>
+						<?php endif; ?>
+
+						<h4 class="jp-relatedposts-post-title">
```
This is no longer relevant since we don't have options the heading in the block itself, in favor of the separate Heading block. See https://github.com/Automattic/wp-calypso/pull/27454 for context and rationale.

tyxla committed:
Exclude previously visited post properly

tyxla committed:
Do not load posts when they have loaded already

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+		);
+
+		$related_posts = $this->get_for_post_id( get_the_ID(), array(
+			'size' => $block_attributes['size'],
```
👍 Done in https://github.com/Automattic/jetpack/pull/10132/commits/09c8b730e6fa1c3b55c4b7cd93442cf66f93e01b

tyxla committed:
Add excerpt to title attribute

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+			'show_date' => isset( $attributes['displayDate'] ) ? (bool) $attributes['displayDate'] : true,
+			'show_context' => isset( $attributes['displayContext'] ) ? (bool) $attributes['displayContext'] : true,
+			'layout' => isset( $attributes['postLayout'] ) && $attributes['postLayout'] === 'list' ? $attributes['postLayout'] : 'grid',
+			'size' => ! empty( $attributes['postsToShow'] ) ? absint( $attributes['postsToShow'] ) : 3,
```
I don't see anything too noisy that would justify adding a bunch of helpers. Unless you have a good idea of how to do it with less code, let's leave this for another time.

tyxla committed:
Prevent the customizer from modifying block output

tyxla committed:
Remove legacy wrapper if we have a Related Posts block in the content

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+		$block_attributes = array(
+			'show_thumbnails' => isset( $attributes['displayThumbnails'] ) ? (bool) $attributes['displayThumbnails'] : false,
+			'show_date' => isset( $attributes['displayDate'] ) ? (bool) $attributes['displayDate'] : true,
+			'show_context' => isset( $attributes['displayContext'] ) ? (bool) $attributes['displayContext'] : true,
```
Seems like you addressed this in https://github.com/Automattic/jetpack/pull/10132/commits/6d9d9dd6a8c721d265890aa59567e6f4d8a204c3 - thanks Jon! 🙌

tyxla committed:
Conditionally add jp-relatedposts-items-visual class

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+						'jp-relatedposts-post' . $index,
+						! empty( $block_attributes['show_thumbnails'] ) ? 'jp-relatedposts-post-thumbs' : '',
+					) );
+					?>
```
After https://github.com/Automattic/jetpack/pull/10132/commits/73b1f29b3c01fa2845b75b14070eaf264cbbabeb this will not happen anymore, because we no longer generate any HTML with JS.

tyxla commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -246,6 +251,118 @@ public function get_target_html_unsupported() {
+		ob_start();
+		?>
+		<div id="jp-relatedposts" class="jp-relatedposts" style="display: block;">
+			<div class="jp-relatedposts-items jp-relatedposts-items-visual jp-relatedposts-<?php echo esc_attr( $block_attributes['layout'] ); ?>">
```
Great catch! Fixed in https://github.com/Automattic/jetpack/pull/10132/commits/82b7dff93db3b06028135f1d5c5de8d6b373021f

tyxla committed:
Some phpcs fixes

tyxla commented:
I have been addressing the feedback today, and I'm nearly done.

> I would also recommend looking at the WPCS reviews for the code you've added; there were a few recommendations you could follow to make sure your code respects the coding standards.

Thanks @jeherve, I have addressed the ones that made sense to me. From what I can see, there are several false positives.

> All in all, is there a reason you did not follow the existing Related Posts implementation that adds the posts via JavaScript?

The main reason was that it uses an alternative "endpoint" for fetching the posts that we won't be able to use outside of Jetpack. That endpoint also doesn't support some of the settings we need to pass to it. So what we're doing instead is, we're using most of the JS for related posts, but fetching the posts and building the output by ourselves.

Also, @sirreal thanks for your thorough feedback, I've addressed your suggestions and took most of your ideas into consideration - they have been really helpful. Seems like we're ending up with limiting the block to 1 per post, and using most of the legacy JS on the front end too, but we work out several conflicts in them manually, and we use the server side for building the output of the related posts.

I'd like to have another fresh look tomorrow and update some testing instructions before I declare this as "ready for review" again.

Thanks everybody, I really appreciate how thorough and detailed your feedback was!

tyxla commented:
This is ready for another review.

lezama commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -248,6 +253,127 @@ public function get_target_html_unsupported() {
+	 */
+	public function render_block( $attributes ) {
+		$block_attributes = array(
+			'show_thumbnails' => isset( $attributes['displayThumbnails'] ) ? (bool) $attributes['displayThumbnails'] : false,
```
```suggestion
			'show_thumbnails' => isset( $attributes['displayThumbnails'] ) && (bool) $attributes['displayThumbnails'],
```

tyxla replied:
Nice, done in https://github.com/Automattic/jetpack/pull/10132/commits/c24f1010b7ba42fc683f81b3159be98310156060.

lezama reviewed (dismissed):
tests :magnifique:

tyxla committed:
Simplify block attribute retrieval

tyxla committed:
Add some spacing

tyxla commented on modules/related-posts/related-posts.js:
```diff
@@ -194,6 +194,13 @@
 
+	function afterPostsHaveLoaded() {
+		jprp.setVisualExcerptHeights();
+		$( '#jp-relatedposts a.jp-relatedposts-post-a' ).click(function() {
```
I left these as they were, because I only moved them. But yeah, since I'm moving them, it doesn't hurt to fix the linting errors. Done in https://github.com/Automattic/jetpack/pull/10132/commits/14cee83bcc4a545331151fedd02176afdbc78b27.

jeherve commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -248,6 +253,127 @@ public function get_target_html_unsupported() {
+			return '';
+		}
+
+		remove_filter( 'the_content', array( $this, 'filter_add_target_to_dom' ), 40 );
```
Instead of using the filter here, I wonder if it would be more readable if we used the same method as for the shortcode:
https://github.com/Automattic/jetpack/blob/14cee83bcc4a545331151fedd02176afdbc78b27/modules/related-posts/jetpack-related-posts.php#L186

tyxla replied:
Nice suggestion 👍 Done in ef92f22f765e756d8fde8ef5d3097c08728377d5.

jeherve commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -248,6 +253,127 @@ public function get_target_html_unsupported() {
+			'size'            => ! empty( $attributes['postsToShow'] ) ? absint( $attributes['postsToShow'] ) : 3,
+		);
+
+		$excludes = array();
```
I wonder if we should move this to its own function to avoid duplicating code?

tyxla replied:
Sure 👍 Done in 0215f1128b6bc68b635b705c7a5e564722475f8c

jeherve commented on modules/related-posts/jetpack-related-posts.php:
```diff
@@ -248,6 +253,127 @@ public function get_target_html_unsupported() {
+
+		ob_start();
+		?>
+		<div id="jp-relatedposts" class="jp-relatedposts jp-relatedposts-block" style="display: block;">
```
This is another area that I wish we could pull out to avoid code duplication, although I realize it would require a lot of changes in the way the existing Related Posts work.

tyxla replied:
Yeah, we'd need to change a bunch, and I preferred not to, as it was going way beyond I'd expect to do for building a block. I realize we introduce some duplication here, but this way we're getting the best of both worlds - reusing as much as possible from the other functionality, while not letting unwanted functionality to stand in our way, but without refactoring a bunch of the existing module.

jeherve reviewed (commented):
This seems to work well! 👍

I have to say that it is still a bit confusing to me in some areas. Here is some feedback (although I realize a lot of it is about the block itself, and not the implementation here).

1. When in the customizer, I have a Related Posts section that I can use to customize the look of all related posts on my site (except for the look of related posts added via a block, since in those cases the block itself decides how the related posts will look like).
But if I add a related posts block to one of my posts, and then view that post in the customizer, the related posts options allow me to customize the look of the block, when in fact when I save my changes, the format of my block won't have changed at all.

2. Related to the above; should we mention that in Settings > Reading and Jetpack > Settings > Traffic? ("The settings below will impact all related posts on your site, except for those you created via the block editor"?)

3. I am seeing a flash of the default related posts in the block in some cases. You can see the default related posts' images and content appear here when I switch from the code editor to the visual editor, before to revert back to my related posts:
https://videopress.com/v/EtzwqZFQ

4. Following up on the above: I've selected the "show thumbnails" option for my block, and yet I do not see any thumbnails for any of the posts that come up. This is a bit confusing. I know that this happens because there are no images in those posts, but I am not sure this will be clear for our users. I wonder if we should grey out the option when no images are found in the related posts?
Somewhat related: #2220 

5. In the existing implementation, we also set `relatedposts_origin` here:
https://github.com/Automattic/jetpack/blob/14cee83bcc4a545331151fedd02176afdbc78b27/modules/related-posts/jetpack-related-posts.php#L148
(we also do some tracking with `_log_click`, but that's only done on WordPress.com)
Should we replicate that in the new implementation?

6. The block inspector offers me the option to display up to 3 posts per block. It would be nice to have the option to display more, especially since that is possible with the existing Related Posts (through a filter).

tyxla committed:
Disable customize controls when content has a related posts block

tyxla committed:
Add disclaimer that blocks are not affected to Settings

tyxla committed:
Add disclaimer that blocks are not affected to Jetpack Traffic settings

tyxla committed:
Support for WP < 5.0

tyxla committed:
Improve how we disable the legacy widget when block is present in content

tyxla committed:
Introduce a parse_numeric_get_arg helper

tyxla commented:
> But if I add a related posts block to one of my posts, and then view that post in the customizer, the related posts options allow me to customize the look of the block, when in fact when I save my changes, the format of my block won't have changed at all.

Good point! I've changed this in 54432449e0adad94af97be71ea595502d616e6d8 - so now, if we have a block in our current post/page, the customizer controls won't be confusingly displayed.

> Related to the above; should we mention that in Settings > Reading and Jetpack > Settings > Traffic? ("The settings below will impact all related posts on your site, except for those you created via the block editor"?)

Sure - makes sense to me. Done in 965aece0c704c8c18439a5621cc7174fdbcc6818 for Settings > Reading and in 84fcfd1e0115e0f6af72fcc1080298a116dbc1e9 for Jetpack > Settings > Traffic - see previews:

![](https://cldup.com/q5RmIoZ7pB.png)

![](https://cldup.com/VIG-pLF7q1.png)

> I am seeing a flash of the default related posts in the block in some cases. You can see the default related posts' images and content appear here when I switch from the code editor to the visual editor, before to revert back to my related posts: https://videopress.com/v/EtzwqZFQ

This should no longer be the case since we landed https://github.com/Automattic/wp-calypso/pull/29596/

> Following up on the above: I've selected the "show thumbnails" option for my block, and yet I do not see any thumbnails for any of the posts that come up. This is a bit confusing. I know that this happens because there are no images in those posts, but I am not sure this will be clear for our users. 

I understand, but this only mirrors what the user will see in the front end, which is a foundational characteristic of how Gutenberg blocks work. If we add placeholder images here, users might be confused too - then they'll have expectations that they'll see those placeholders in the frontend, and they won't.

> I wonder if we should grey out the option when no images are found in the related posts?

Well, I say we shouldn't, because this could potentially be even more confusing. What if the user decides to add featured images to those posts later? I say let's leave this decision in the hands of the user.

> In the existing implementation, we also set relatedposts_origin here:
> (we also do some tracking with _log_click, but that's only done on WordPress.com)
> Should we replicate that in the new implementation?

I'm not sure I'm following you on this one - could you please elaborate? I'm not sure why and where in our implementation we'd need it.

> The block inspector offers me the option to display up to 3 posts per block. It would be nice to have the option to display more, especially since that is possible with the existing Related Posts (through a filter).

This doesn't seem like a blocker to me - it would be nice to have, but I think you'll agree we can address it in a subsequent PR.

tyxla commented:
Thanks for the thorough feedback! I've addressed it, left some comments, and this is ready for another 👀

tyxla committed:
Fixup

jeherve reviewed (dismissed):
This works well now. I only have a few last remarks, but nothing that can't be addressed in a new PR in the future. 👍

tyxla commented:
Rebased to fix a conflict in `_inc/client/traffic/related-posts.jsx`. Can we get another ✅ ?

jeherve reviewed (approved):
Approving again, for good measure

tyxla commented:
Thanks @jeherve! We'll address the minor feedback in subsequent PRs. Merging this one now.
